aten/src/ATen/core/ivalue.h

```diff
@@ -473,12 +480,15 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
 
   // Get the result of the current future.
   IValue value() {
+    std::unique_lock<std::mutex> lock(mutex_);
     AT_ASSERT(completed());
     return value_;
   }
 
   void addCallback(std::function<void(void)> callback) {
+    std::unique_lock<std::mutex> lock(mutex_);
     if (completed()) {
+      lock.unlock();
```

Comment:
this seems like it is missing a return. It is adding callback to the callbacks regardless. This will leak memory because the future will hold the closed state in the callback.

Reply:
Good catch

Reply:
why don't we add the callback to the callbacks? are we going to reuse the future (e.g. for perf) or is it only used once?

Reply:
because a `callback` is saved in `callbacks` until the future completes. If the future has already completed, there is no need to save it; we can just directly invoke it.

Reply:
right, I assume the future is a single-use future?

Reply:
A future can be waited by multiple instructions like
```lang=python
future = jit._fork(foo)
a = jit._wait(future)
b = jit._wait(future)
assert a == b
```
And yes for now, only a single thread can wait on a future.